Index: policykit-1-0.105.10.5/src/polkitagent/polkitagenthelper-pam.c
===================================================================
--- policykit-1-0.105.10.5.orig/src/polkitagent/polkitagenthelper-pam.c
+++ policykit-1-0.105.10.5/src/polkitagent/polkitagenthelper-pam.c
@@ -32,6 +32,7 @@
 #include <security/pam_appl.h>
 
 #include <polkit/polkit.h>
+#include <stdbool.h>
 
 static int conversation_function (int n, const struct pam_message **msg, struct pam_response **resp, void *data);
 
@@ -70,6 +71,37 @@ send_to_helper (const gchar *str1,
   g_free (tmp2);
 }
 
+void load_default_locale() {
+  char buff[512] = "\0";
+
+  FILE *f = fopen("/etc/default/locale", "r");
+  char *pos;
+  char *value;
+
+  if (f == NULL) {
+  return;
+  }
+  while (fgets(buff, 512, f) != NULL) {
+  if ((pos = strchr(buff, '\n')) != NULL) {
+      *pos = '\0';
+  }
+
+  pos = strchr(buff, '=');
+  if (pos == NULL || buff == pos) {
+      continue;
+  }
+
+  *pos = '\0';
+
+  char *value = pos + 1;
+#ifdef PAH_DEBUG
+  fprintf (stdout, "polkit-agent-helper-1: set `%s': %s\n", buff, value);
+#endif /* PAH_DEBUG */
+  setenv(buff, value, true);
+  }
+  fclose(f);
+}
+
 int
 main (int argc, char *argv[])
 {
@@ -131,6 +163,8 @@ main (int argc, char *argv[])
         }
     }
 
+  load_default_locale();
+
 #ifdef PAH_DEBUG
   fprintf (stderr, "polkit-agent-helper-1: user to auth is '%s'.\n", user_to_auth);
 #endif /* PAH_DEBUG */
