Index: b/src/polkitagent/polkitagenthelper-deepin.c
===================================================================
--- a/src/polkitagent/polkitagenthelper-deepin.c
+++ b/src/polkitagent/polkitagenthelper-deepin.c
@@ -513,9 +513,8 @@ void sig_hanlder(struct sigaction *act)
     if (_pda)
     {
         _pda->exit = true;
-        if (_pda->loop)
+        if (_pda->loop && g_main_loop_is_running(_pda->loop))
         {
-	    flush_and_wait();
             g_main_loop_quit(_pda->loop);
         }
     }
