Index: policykit-1-0.105.9.10+c1/src/programs/pkexec.c
===================================================================
--- policykit-1-0.105.9.10+c1.orig/src/programs/pkexec.c
+++ policykit-1-0.105.9.10+c1/src/programs/pkexec.c
@@ -60,6 +60,7 @@ static struct passwd *pw;
 
 #define MAX_BUF_SIZE 1024
 #define DEVELOPER_MODE_FILE "/var/lib/deepin/developer-mode/enabled"
+#define DEVELOPER_MODE_DEV_FILE "/dev/elf_verifier"
 #define DEVELOPER_MODE_ENABLE_STR "1"
 #define SECURITY_WHILTELIST_FILE "/var/lib/deepin/deepin_security_verify.whitelist"
 
@@ -434,7 +435,9 @@ security_verify_check(const char *path)
     if (has_in_developer_mode()) {
         return 1;
     }
-
+    if (access(DEVELOPER_MODE_DEV_FILE, F_OK)){
+        return 1;
+    }
     return has_in_whitelist(path);
 }
 
