Index: b/src/polkitagent/polkitagenthelper-deepin.c
===================================================================
--- a/src/polkitagent/polkitagenthelper-deepin.c
+++ b/src/polkitagent/polkitagenthelper-deepin.c
@@ -572,6 +572,17 @@ int main(int argc, char *argv[])
         goto error;
     }
 
+    load_default_locale();
+
+    char *lang = NULL;
+    if (!da_get_user_lang(argv[1], &lang)) {
+        if (lang) {
+            helper_log("set env LANG = %s", lang);
+            setenv("LANG", lang, 1);
+            free(lang);
+        }
+    }
+
     // 注册 SIGTERM 回调
     if (signal(SIGTERM, sig_hanlder) == SIG_ERR)
     {
