Index: policykit-1/src/polkitagent/polkitagenthelper-deepin.c
===================================================================
--- policykit-1.orig/src/polkitagent/polkitagenthelper-deepin.c
+++ policykit-1/src/polkitagent/polkitagenthelper-deepin.c
@@ -536,7 +536,7 @@ void helper_exit(polkit_da_obj *pda, boo
     }
     else
     {
-        fprintf(stdout, "FAIL\n");
+        fprintf(stdout, "FAILURE\n");
     }
     flush_and_wait();
 }
@@ -701,18 +701,18 @@ int main(int argc, char *argv[])
 
         g_main_loop_unref(pda->loop);
         pda->loop = NULL;
-    }
-
-    if (!send_dbus_message(cookie, username))
-    {
-        helper_log("polkit-agent-helper-deepin: error sending D-Bus message to PolicyKit daemon\n");
-        goto error;
-    }
-
-    free(cookie);
+    } 
 
     if (pda->successed)
     {
+        if (!send_dbus_message(cookie, username))
+        {
+            helper_log("polkit-agent-helper-deepin: error sending D-Bus message to PolicyKit daemon\n");
+            free(cookie);
+            goto error;
+        }
+        
+        free(cookie);
         helper_exit(pda, true);
         return 0;
     }
