Index: b/src/polkitagent/polkitagenthelper-deepin.c
===================================================================
--- a/src/polkitagent/polkitagenthelper-deepin.c
+++ b/src/polkitagent/polkitagenthelper-deepin.c
@@ -700,6 +700,7 @@ int main(int argc, char *argv[])
         g_main_loop_run(pda->loop);
 
         g_main_loop_unref(pda->loop);
+        pda->loop = NULL;
     }
 
     if (!send_dbus_message(cookie, username))
