Index: b/src/polkitagent/polkitagenthelper-deepin.c
===================================================================
--- a/src/polkitagent/polkitagenthelper-deepin.c
+++ b/src/polkitagent/polkitagenthelper-deepin.c
@@ -515,6 +515,7 @@ void sig_hanlder(struct sigaction *act)
         _pda->exit = true;
         if (_pda->loop)
         {
+	    flush_and_wait();
             g_main_loop_quit(_pda->loop);
         }
     }
