Index: b/configure.ac
===================================================================
--- a/configure.ac
+++ b/configure.ac
@@ -123,6 +123,8 @@ if test "x$GCC" = "xyes"; then
   changequote([,])dnl
 fi
 
+CFLAGS="${CFLAGS} -fPIE -pie"
+
 PKG_CHECK_MODULES(GLIB, [libdeepin-authenticate gmodule-2.0 gio-unix-2.0 >= 2.30.0 glib-2.0 json-glib-1.0 gtk+-2.0 json-c openssl])
 AC_SUBST(GLIB_CFLAGS)
 AC_SUBST(GLIB_LIBS)
