Index: b/src/polkitagent/polkitagenttextlistener.c
===================================================================
--- a/src/polkitagent/polkitagenttextlistener.c
+++ b/src/polkitagent/polkitagenttextlistener.c
@@ -204,6 +204,7 @@ on_completed (PolkitAgentSession *sessio
 {
   PolkitAgentTextListener *listener = POLKIT_AGENT_TEXT_LISTENER (user_data);
 
+  polkit_agent_session_ctrl(session, -1, SESSION_CTRL_CLOSE);
   if (listener->use_color)
     fprintf (listener->tty, "\x1B[1;31m");
   if (gained_authorization)
@@ -536,6 +537,7 @@ polkit_agent_text_listener_initiate_auth
                                                NULL);
 
   polkit_agent_session_initiate (listener->active_session);
+  polkit_agent_session_ctrl(listener->active_session, -1, SESSION_CTRL_OPEN);
 
  out:
   ;
