Index: b/src/polkitagent/polkitagenthelper-deepin.c
===================================================================
--- a/src/polkitagent/polkitagenthelper-deepin.c
+++ b/src/polkitagent/polkitagenthelper-deepin.c
@@ -477,6 +477,7 @@ static gboolean send_auth_info_to_helper
     da_prop_get_is_MFA(pda->da, &is_mfa);
 
     pda->had->auth_info.is_mfa = is_mfa;
+    pda->had->auth_info.auth_flags = NULL;
 
     da_factor_info *dfi = NULL;
     int factor_num = 0;
