Index: policykit-1/src/polkit/polkitidentity.c
===================================================================
--- policykit-1.orig/src/polkit/polkitidentity.c
+++ policykit-1/src/polkit/polkitidentity.c
@@ -163,7 +163,7 @@ polkit_identity_from_string  (const gcha
       val = g_ascii_strtoull (str + sizeof "unix-user:" - 1,
                               &endptr,
                               10);
-      if (*endptr == '\0')
+      if (*endptr == '\0' && getpwuid((uid_t) val))
         identity = polkit_unix_user_new ((gint) val);
       else
         identity = polkit_unix_user_new_for_name (str + sizeof "unix-user:" - 1,
